### PR TITLE
Support vertex offset mode with new D3D12_VERTEX_BUFFER_VIEW

### PR DESCRIPTION
**Pull request by AMD-dwang**
In vertex offset mode. we will use D3D12_VERTEX_BUFFER_VIEW to replace SRD, the address and size fields are already in the right place. We only need to get stride and fill the fourth DWORD. Then we calculate offset with stride and vertex index.

For Uber fetch shader, SRD will be updated in driver.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7857706f4afda373c97e3678eccb7066e4875a5c
<details><summary>CTS tests (Failed: 1235/138378)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35162/69163 (50.8%)</li>
<li>Failed: 0/69163 (0.0%)</li>
<li>Not Supported: 34001/69163 (49.2%)</li><li>Warnings: 0/69163 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 34018/69215 (49.1%)</li>
<li><details><summary>Failed: 1235/69215 (1.8%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer.suballocation.storage.index.indirect.create.binding
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_dst.vertex.indirect.create.binding_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_src.uniform_texel.vertex.create.binding_residency
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.vertex.indirect.create.binding_residency
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.vertex.indirect.create.binding_residency_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer_marker.graphics.external_host_mem.bottom_of_pipe.memory_dep.draw
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer_view.access.uniform_texel_buffer.a2b10g10r10_uint_pack32
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer_view.access.uniform_texel_buffer.r32_sfloat
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.color.3d.a8b8g8r8_sint_pack32.a8b8g8r8_sint_pack32.optimal_optimal_nearest_stripes_x
Stack trace: Fail

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.color.3d.r16g16b16a16_snorm.r16g16b16a16_snorm.optimal_general_linear_stripes_x
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.depth_stencil_msaa_copy.whole.d16_unorm_s8_uint_general_general_D_2_bit
Stack trace: Fail

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.depth_stencil_msaa_copy.whole.d16_unorm_s8_uint_general_optimal_D_4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.diff_image_size.dst_266_256_1_4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.with_regions.2_bit
Stack trace: Fail

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.simple_tests.scaling_whole2_3d.r32_sfloat_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.simple_tests.whole_3d.b8g8r8a8_unorm_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.buffer_to_buffer_with_offset.3_0
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.buffer_to_depthstencil.d32_sfloat_s8_uint_S
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.depth_stencil_msaa_copy.partial.s8_uint_S_4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.depth_stencil_msaa_copy.partial.s8_uint_S_8_bit
Stack trace: Flake
...
```
</p></details></li>

<li>Not Supported: 33962/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Thanks! Let's try to find a better name for this in the compiler interface. As-is, the name is misleading because as far as the contract between driver and compiler is concerned, the point of the option isn't whether `idxen` is used or not. The point of the option is the structure of "vertex buffer descriptors" in memory.

Even with vertex buffer descriptors laid out in the alternative way, the compiler could decide to set up a HW buffer descriptor with the stride and use `idxen` instructions. On current HW, this would require dividing the buffer size by the stride (and it's unclear what should happen if the size is not a multiple of the stride), so the implementation path you took is better. But this may change in future HW.

How about `useSoftwareVertexBufferDescriptors` as an alternative name for the option? The idea being that the VB descriptors are laid out in a way that is native for software, as opposed to the layout that is native for HW. But I'm open to alternative suggestions. (I briefly considered that we could use an enum, but that seems overkill.)

Apart from that, I believe all uses of `amdgcn_struct_*buffer_load*` intrinsics should be replaced by `amdgcn_raw_*buffer_load*` intrinsics when in this mode. Otherwise, we may needlessly pass in a VGPR with a value of 0. This requires more changes throughout `fetchVertex` and `addVertexFetchInst` (but mostly it is a matter of adjusting `args` and making the intrinsic ID dynamic).

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7857706f4afda373c97e3678eccb7066e4875a5c
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by AMD-dwang** [comment]
V1: address the comments. 
 Rename the name for the option. And replace the intrinsics.

File: lgc/patch/VertexFetch.cpp [hunk 1]
```diff
@@ -879,22 +899,14 @@ Value *VertexFetchImpl::fetchVertex(InputImportGenericOp *inst, Value *descPtr,
...
   Value *wholeVertex = nullptr;
   {
     builder.SetInsertPoint(wholeVertexBlock);
-    wholeVertex = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+    wholeVertex = builder.CreateIntrinsic(instId, fetchType, args, {});
     if (is64bitFetch) {
       // If it is 64-bit, we need the second fetch
-      args[2] = builder.CreateAdd(args[2], builder.getInt32(SizeOfVec4));
-      auto secondFetch = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+      args[offsetIdx] = builder.CreateAdd(args[2], builder.getInt32(SizeOfVec4));
```
Comment (nhaehnle) [inline lgc/patch/VertexFetch.cpp:908]: args[2] -> args[offsetIdx]

**Review by nhaehnle — APPROVED** [review]
Thank you, and thank you for cleaning up the code as you went along (e.g. using builder methods). Looking really good!

File: lgc/patch/VertexFetch.cpp (hunk 1 above)
Comment (AMD-dwang) [reply lgc/patch/VertexFetch.cpp]: Fixed. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 404c2ced19825388427859a117dbae6e5dc595de
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
Thanks!

**Pull request merged**